cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/ObjectMappedTable.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.api.dataset.lib;
+
+import co.cask.cdap.api.annotation.Beta;
+import co.cask.cdap.api.data.batch.BatchReadable;
+import co.cask.cdap.api.data.batch.BatchWritable;
+import co.cask.cdap.api.data.batch.RecordScannable;
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.cdap.api.dataset.Dataset;
+
+import javax.annotation.Nullable;
+
+/**
+ * A Dataset that stores plain java Objects into a table by mapping object fields to table columns.
+ * Objects must be flat and can only contain fields of simple type. A simple type is an Integer,
+ * int, Long, long, Float, float, Double, double, String, byte[], ByteBuffer, or UUID.
+ * The object itself cannot be a simple type.
+ *
+ * @param <T> the type of objects in the table
+ */
+@Beta
+public interface ObjectMappedTable<T> extends Dataset, BatchReadable<byte[], T>,
+  BatchWritable<byte[], T>, RecordScannable<StructuredRecord> {
+
+  /**
+   * Write an object with a given key.
+   *
+   * @param key the key of the object
+   * @param object the object to be stored
+   */
+  void write(String key, T object);
+
+  /**
+   * Write an object with a given key.
+   *
+   * @param key the key of the object
+   * @param object the object to be stored
+   */
+  void write(byte[] key, T object);
+
+  /**
+   * Read an object with a given key.
+   *
+   * @param key the key of the object
+   * @return the object if found, or null if not found
+   */
+  T read(String key);
+
+  /**
+   * Read an object with a given key.
+   *
+   * @param key the key of the object
+   * @return the object if found, or null if not found
+   */
+  T read(byte[] key);
+
+  /**
+   * Scans table.
+   *
+   * @param startRow start row inclusive. {@code null} means start from first row of the table
+   * @param stopRow stop row exclusive. {@code null} means scan all rows to the end of the table
+   * @return {@link CloseableIterator} over {@link KeyValue KeyValue&lt;byte[], T&gt;}
+   */
+  CloseableIterator<KeyValue<byte[], T>> scan(@Nullable String startRow, @Nullable String stopRow);
```

Comment:
As a side thought: This seems right as KeyValueTable returns a CloseableIterator, too. I am just wondering why Table is returning a Scanner? Should we change Table?